feat: #6 회원가입 (member)

tpleehan opened the pull request:
# #6 회원가입 구현
## Description :memo:
- 회원가입 기능을 구현했습니다.
- 민감하고 개인정보에 관련된 데이터를 plant text 또는 raw 데이터로 데이터베이스에 저장하는 것이 아닌
단방향 암호화 SHA512를 적용 했습니다. (password) 
- Email JWT 인증 기능을 구현했습니다.

## To Reviewers :raising_hand:
- JPA 학습 후 적용하고 있는데, 예전에 학습하던 방식과 다르지만 확실히 애플리케이션 데이터를 객체 지향 관점에서 바라 볼 수 있어서 학구열을 많이 느끼게 되었습니다. 😃
- 테스트 코드가 익숙하지 않아서 미흡하고 부족합니다. (여러 방면의 테스트를 하기 위한 공부가 필요하다고 느꼈습니다. 😭)
- 코드의 흐름과 옳고 그름에 대한 판단이 많이 부족하여 리뷰를 통해 좋은 부분들을 흡수하도록 하겠습니다. !! :blush:

tpleehan committed:
feat: #2 Project init setting

프로젝트 초기 설정
- Coding Convetion / naver hackday conventions 설정
- Restful API 문서화 도구 / Swagger 설정

tpleehan committed:
build: #2 Project init setting - Swagger 및 gitignore 추가

프로젝트 초기 설정
- Swagger / springfox -> springdoc 변경(최신 버전 적용)
- application.properties 설정 파일 gitignore 추가

tpleehan committed:
feat: #2 Project init setting - Swagger 및 gitignore 추가

프로젝트 초기 설정
- Swagger / springfox -> springdoc 변경(최신 버전 적용)
- application.properties 설정 파일 gitignore 추가

tpleehan committed:
Merge remote-tracking branch 'origin/feature/2-project-init-setting' into feature/2-project-init-setting

tpleehan committed:
feat: #6 회원가입 (member)

- 일반 회원 회원가입 기능 구현

tpleehan committed:
feat: #6 회원가입 (member) 기능 구현

- lombok 추가
- 일반 회원 회원가입 기능 구현

tpleehan committed:
fix: #6 회원가입 (member) 기능 구현 중 에러 발생 확인

- http 확인 중 500 error로 Dto, Service, Controller 수정

tpleehan committed:
feat: #6 회원가입 (member) 기능 구현

- 암호화 적용 (SHA512 Encryption 적용)
- Email JWT 인증 적용

tpleehan committed:
chore: properties -> yml로 변경

- application.properties 속성 파일을 application.yml로 변경

tpleehan committed:
chore: gitignore 수정

- application.yml 추가

tpleehan committed:
chore: properties -> yml로 변경

- application.properties 속성 파일을 application.yml로 변경

tpleehan committed:
Revert "chore: properties -> yml로 변경"

This reverts commit 27ac812376f7b0a53c55c74a008da265889b2398.

brody-0125 commented on src/main/java/com/project/lala/common/encrytion/SHA512EncryptionService.java:
```diff
@@ -0,0 +1,30 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class SHA512EncryptionService implements EncryptionService {
```
암호화 로직을 스프링의 관리를 받는 Bean으로 등록하여 사용하신 이유가 궁금하네요! 

객체를 Bean를 등록하면 어떠한 이점이 있을까요?

tpleehan replied:
객체를 Bean으로 등록하면 스프링이 스프링 컨테이너에 객체에 대한 생성 및 생명주기를 관리할 수 있도록 위임하여 여러 클래스에 동일한 객체를 여러 개 만들 필요 없는 이점이 있습니다.

brody-0125 commented on src/main/java/com/project/lala/common/encrytion/SHA512EncryptionService.java:
```diff
@@ -0,0 +1,30 @@
+			MessageDigest messageDigest = MessageDigest.getInstance("SHA-512");
+
+			byte[] rawBytes = raw.getBytes();
+			messageDigest.reset();
```
MessageDigest를 사용하기 전에 `messageDigest.reset()` 을 호출하는 이유는 무엇인가요?

tpleehan replied:
`messageDigest.reset()`를 호출한 이유는 해당 인스턴스를 재사용하기 위해서 digest를 초기화 하려고 호출하게 되었습니다. 값을 해싱하기 전 해싱 인스턴스를 초기화를 위함입니다.

brody-0125 commented on src/main/java/com/project/lala/common/encrytion/SHA512EncryptionService.java:
```diff
@@ -0,0 +1,30 @@
+			}
+			return sb.toString();
+		} catch (NoSuchAlgorithmException e) {
+			return raw;
```
NoSuchAlgorithmException은 `MessageDigest.getInstance()`에 잘못된 Algorithm을 넘겨줬을 때 발생하는 Exception이고, 테스트를 수행했다면 발생할 여지는 없는데요.

별도의 로그를 남기지 않고 입력 값을 다시 돌려주는 이유는 무엇인가요?

tpleehan replied:
로직 작성 시 해당 부분에 대해 고려를 하지 않고 로직을 작성하게 되었습니다. 로직 수정을 하겠습니다!!

brody-0125 commented on .gitignore:
```diff
@@ -35,3 +35,6 @@ out/
 .vscode/
+
+### application.yml ###
+application.yml
```
대부분의 파일에서 No newline at a end of file이 발생하고 있는 것 같아요.

이와 관련된 설정이나 Formatter가 제대로 적용되어 있는지 확인 부탁드립니다~

tpleehan replied:
git commit 시 체크를 하지 않고 push를 해버렸습니다…😥
Formatter 설정과 함께 CheckStyle을 이용해서 꼭 확인하겠습니다!!

brody-0125 commented on src/main/java/com/project/lala/common/member/MemberStatus.java:
```diff
@@ -0,0 +1,11 @@
+ */
+public enum MemberStatus {
+	UNEMAILAUTH, EMAILAUTH, AUTH, DELETED
+}
```
Enum이나 Constant를 정의해주실 때에는 단어 사이에 `_`를 사용하는 것이 Java Officail Convention입니다~

brody-0125 replied:
현재 Member에 대한 Authorization 이 Email로만 수행되고 있긴 하지만, 추후의 다른 방식을 허용한다면 해당 Naming외에도 추가적인 것들이 생겨날 여지가 있을 것 같아요~

tpleehan replied:
Enum과 Constant에 대한 Java Convention을 다시 숙지하도록 하겠습니다!

brody-0125 commented on src/main/java/com/project/lala/controller/MemberController.java:
```diff
@@ -0,0 +1,37 @@
+import lombok.RequiredArgsConstructor;
+import lombok.extern.log4j.Log4j2;
+
+@Log4j2
```
`@Log4j2`를 이용해 Logger를 등록하신 이유는 무엇인가요?

혹시 slf4j에 대해서 알고 계신가요? 
slf4j를 사용하는 것과 log4j2를 사용하는 것의 차이를 정리해와 주시면 좋을 것 같습니다.

tpleehan replied:
`@Slf4j`를 명시해야 하는데 `@Log4j2`로 명시하게 되어 수정하겠습니다! 😓

tpleehan replied:
정리한 부분에 대해 링크 공유합니다!!
https://leehan-project.notion.site/slf4j-log4j2-131e7ceef2f0470a9d951a79fac1f1bb

brody-0125 commented on src/main/java/com/project/lala/controller/MemberController.java:
```diff
@@ -0,0 +1,37 @@
+	private final MemberService memberService;
+	private final EmailService emailService;
+
+	@PostMapping("/signup")
```
새로운 Member를 생성하는 API에 `/signup`을 사용하신 이유는 어떤 것일까요?  

RESTful 규약 상 Member라는 요소를 어떤 것으로 규정하고 어떻게 URL을 작성하도록 요구하는지 확인해보시면 좋을 것 같습니다~

_물론 이러한 URL 사용할만한 애매한 상황이 존재하기도 합니다._

tpleehan replied:
MemberController에는 Member에 관련된 부분만 로직이 구성되어야 하는데 따로 고려하지 않고 해당 API를 사용하게 되었습니다. 해당 부분은 따로 수정하겠습니다~!

brody-0125 commented on src/main/java/com/project/lala/controller/MemberController.java:
```diff
@@ -0,0 +1,37 @@
+	private final EmailService emailService;
+
+	@PostMapping("/signup")
+	public void signup(@RequestBody SignUpRequestDto signUpRequestDto) {
```
Request와 Dto라는 Suffix는 사실 의미가 중복되는 느낌이 있습니다.

Dto는 통신 용도의 객체를 의미하고, Request 또한 Client의 요청을 변환한 객체를 의미하거든요~ 어떻게 정리하면 좋을지 고민해보시면 좋을 것 같네요.

brody-0125 replied:
또한 해당 객체에 대한 Validation이 적용되어 있지 않은 것 같네요.

tpleehan replied:
SignUpRequestDto를 SignUpRequest로 변경을 하는 것이 옳을까요?

brody-0125 commented on src/main/java/com/project/lala/dto/SignUpRequestDto.java:
```diff
@@ -0,0 +1,61 @@
+public class SignUpRequestDto {
+
+	@Column(name = "login_id")
+	@Size(min = 6, max = 30, message = "아이디는 최소 6자 이상 이어야 합니다.")
```
통신 객체에 `javax.persistence.Column` 가 사용되는 이유는 무엇인가요?

tpleehan replied:
해당 부분은 수정하도록 하겠습니다 !

brody-0125 commented on src/main/java/com/project/lala/dto/SignUpRequestDto.java:
```diff
@@ -0,0 +1,61 @@
+		this.name = name;
+		this.email = email;
+		this.authToken = authToken;
+	}
```
해당 객체에서 `@Builder`를 사용하고 있는 이유가 무엇인지 궁금하네요.

tpleehan replied:
`@Builder`에 대해 제대로 숙지를 하지 않고 사용하게 된 거 같습니다. 해당 부분은 수정하겠습니다!

brody-0125 commented on src/main/java/com/project/lala/dto/SignUpResponseDto.java:
```diff
@@ -0,0 +1,45 @@
+
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class SignUpResponseDto {
```
해당 객체도 RequestDto와 같이 의미가 중복되는 것으로 보이는데 어떻게 정리하면 좋을지 고민해보시면 좋을 것 같아요.

tpleehan replied:
해당 부분도 SignUpResponse로 변경하고자 합니다!

brody-0125 commented on src/main/java/com/project/lala/entity/EmailAuth.java:
```diff
@@ -0,0 +1,50 @@
+@Getter
+@NoArgsConstructor(access = AccessLevel.PROTECTED)
+public class EmailAuth {
+	private static final long EMAIL_TOKEN_EXPIRATION_TIME_VALUE = 5L;
```
`EMAIL_TOKEN_EXPIRATION_TIME_VALUE` 는 어떤 시간 단위(mills, sec, min..)로 설정되는지 사용되는 지점을 보기 전까지 알기 어려운 것 같은데요~ 

어떻게 하면 상수 이름으로 그런 시간 단위까지 나타낼 수 있을지 고민해보시면 좋을 것 같아요.

또한 변경될 여지가 없는 값일까요? 정책이 바뀔 가능성이 있다면 어떻게 관리되어야 할까요?

tpleehan replied:
시간 단위에 대한 의미를 상수 이름에 부여하면 될까요? 정책이나 변경될 여지는 있는데 해당 부부분에 대해 조금 더 고민해보겠습니다.

brody-0125 commented on src/main/java/com/project/lala/entity/EmailAuth.java:
```diff
@@ -0,0 +1,50 @@
+	private static final long EMAIL_TOKEN_EXPIRATION_TIME_VALUE = 5L;
+
+	@Id
+	@GeneratedValue(strategy = GenerationType.IDENTITY)
```
`GenerationType.IDENTITY`와 다른 GenerationType의 차이는 무엇이 있을지 정리해보시면 좋을 것 같네요!

brody-0125 commented on src/main/java/com/project/lala/entity/EmailAuth.java:
```diff
@@ -0,0 +1,50 @@
+	private boolean expired;
+
+	@Column(name = "expired_at")
+	private LocalDateTime expiredAt;
```
JPA에서 Entity Column을 생성할 때 어떤 Naming Convention을 기본 제공하는지 확인해보시면 좋을 것 같아요~

brody-0125 commented on src/main/java/com/project/lala/entity/EmailAuth.java:
```diff
@@ -0,0 +1,50 @@
+		this.expiredAt = LocalDateTime.now().plusMinutes(EMAIL_TOKEN_EXPIRATION_TIME_VALUE);
+	}
+
+	public void useToken() {
```
함수 명은 `useToken()` 인데 expired 속성을 true로 수정하고 계시네요. expired가 시간에 의한 만료 외에도 다른 의미를 내포하고 있는 것 같은데요.

사용과 시간에 의한 만료가 구분 되어야할 필요가 있을지 고민해보시면 좋을 것 같아요!

brody-0125 commented on src/main/java/com/project/lala/entity/Member.java:
```diff
@@ -0,0 +1,111 @@
+
+	@ElementCollection(fetch = FetchType.LAZY)
+	@Enumerated(EnumType.STRING)
+	private List<MemberStatus> memberStatuses = new ArrayList<>();
```
Memer의 상태가 왜 List 형태로 관리되고 있을까요? 
왜 `@ElementCollection`을 사용하신 것인지 궁금하네요.

tpleehan replied:
해당 부분은 수정하겠습니다!

brody-0125 commented on src/main/java/com/project/lala/entity/Member.java:
```diff
@@ -0,0 +1,111 @@
+		member.email = email;
+		member.registeredAt = LocalDate.now();
+		return member;
+	}
```
Builder를 내부적으로 정의하고 계신데 Factory Method에서 사용하지 않으신 이유는 무엇인가요? 
또한 setter 방식을 사용했을 때 발생할만한 문제점은 무엇이 있을까요?

tpleehan replied:
Factory Method에 Builder를 사용하지 않은 부분에 대한 부분은 숙지가 안 되었던 거 같습니다.😥
`setter` 방식을 사용하게 된다면 혼자 개발하면 상관 없지만, 다른 개발자와 협력하여 프로젝트를 진행할 경우 악의적인(?) 목적으로 값을 변경할 수 있을 뿐더러, 변경하면 안 되는 중요한 값이더라도 변경할 수 있는 값으로 파악되기 때문에 setter의 사용 목적과 의도를 파악하기가 힘듭니다.

참고
https://ssdragon.tistory.com/78

brody-0125 commented on src/main/java/com/project/lala/repository/MemberRepository.java:
```diff
@@ -0,0 +1,17 @@
+import com.project.lala.entity.Member;
+
+@Repository
+public interface MemberRepository extends JpaRepository<Member, Long> {
```
JpaRepository Interface를 구현한 Repository에서 `@Repository`를 사용하고 계신 이유는 무엇인가요?

추가적으로 `@Repository`는 어떨 때 사용하고, 어떤 기능을 제공해줄까요?

tpleehan replied:
`@Repository` 어노테이션은 꼭 명시해야 되는 것으로 알고 있었는데 인터페이스 부분에선 명시하지 않아도 되는걸 알게 되었습니다.
https://stackoverflow.com/questions/44069367/repository-not-necessary-when-implementing-jparepository

`@Repository`는 DB에 접근하는 메서드를 사용하기 위한 인터페이스입니다. 주로 CRUD 등 정의해주는 계층이라 볼 수 있습니다.

brody-0125 commented on src/main/java/com/project/lala/service/EmailService.java:
```diff
@@ -0,0 +1,28 @@
+		mailMessage.setTo(email);
+		mailMessage.setSubject("회원가입 이메일 인증");
+		mailMessage.setText(
+			"http://localhost:8080/api/members/confirm-email?email=" + email + "&authToken=" + authToken);
```
`http://localhost:8080` 이 값이 바뀔 경우는 없을까요? 다른 환경에 배포되어야 한다면 제대로 동작할 수 있을지 고민해보시면 좋을 것 같아요.

+ 외부에 변수를 두고 관리하는 방법에는 무엇이 있고 어떻게 참조할 수 있을지? 찾아보세요.

tpleehan replied:
다른 환경에서 배포할 때 관리할 수 있는 방법으로 yml이나 properties 등을 통해 설정해보겠습니다~!

brody-0125 commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -0,0 +1,85 @@
+@Log4j2
+@Service
+@RequiredArgsConstructor
+@Transactional
```
서비스 Type Level에 `@Transactional` 적용하신 이유는 무엇인가요?

`@Transactional`이 모든 상황에서 필요한 기능일지 정리해 와주시면 좋을 것 같아요,

tpleehan replied:
Service Layer에서 DB와 non-DB 관련 서비스를 모두 가질 수 있기 때문에 `@Transactional`을 사용하게 되었습니다.  클래스, 메서드 레벨에서 사용될 수 있는데, 메서드 부분에 `@Transactional`을 작성 했는데 클래스 레벨에도 적용을 하게 된 부분을 수정하겠습니다!

참고
https://vladmihalcea.com/spring-transactional-annotation/
https://goodteacher.tistory.com/252

Lob-dev commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -0,0 +1,85 @@
+				.build());
+
+		emailService.sendEmail(emailAuth.getEmail(), emailAuth.getAuthToken());
+
+		return SignUpResponseDto.builder()
+			.id(member.getId())
+			.loginId(member.getLoginId())
+			.password(member.getPassword())
+			.nickname(member.getNickname())
+			.name(member.getName())
+			.email(member.getEmail())
+			.build();
```
```
public static SignUpResponseDto signUpResponseDto(Member member) {
		return SignUpResponseDto.builder()
			.id(member.getId())
			.loginId(member.getLoginId())
			.password(member.getPassword())
			.nickname(member.getNickname())
			.name(member.getName())
			.email(member.getEmail())
			.build();
	}
```

통신 객체 안에 해당 로직이 있었는데요. 따로 Builder를 쓰신 이유는 무엇인가요?

brody-0125 replied:
```
public static SignUpResponseDto signUpResponseDto(Member member) {
		return SignUpResponseDto.builder()
			.id(member.getId())
			.loginId(member.getLoginId())
			.password(member.getPassword())
			.nickname(member.getNickname())
			.name(member.getName())
			.email(member.getEmail())
			.build();
	}
```

통신 객체 안에 해당 로직이 있었는데요. 따로 Builder를 쓰신 이유는 무엇인가요?

brody-0125 commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -0,0 +1,85 @@
+	private final EmailAuthRepository emailAuthRepository;
+
+	private final EmailService emailService;
+	private final EncryptionService encryptionService;
```
Domain Service에서 다른 Domain Service를 참조하고 있는 경우 무슨 문제가 생길 수 있을까요? 문제와 다른 방법을 고민해보시면 좋을 것 같아요.

tpleehan replied:
계층 구조를 전혀 고려하지 않고 코드를 작성하게 되었네요😖 Service에서 다른 Service를 참조할 수는 있기 때문에 막연하게 작성한 부분인데, 순환참조를 방지하기 위해서 다른 방법을 고려해야겠습니다!!

참고
https://www.inflearn.com/questions/111740/comment/84542
https://taesan94.tistory.com/268
https://velog.io/@sumusb/Spring-Service-Layer에-대한-고찰

brody-0125 commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -0,0 +1,85 @@
+
+	private void validateDuplicateByLoginId(String loginId) {
+		if (memberRepository.findByLoginId(loginId).isPresent()) {
+			throw new IllegalStateException("이미 존재하는 아이디 입니다.");
```
해당 Exception이 발생하면 프론트 개발자나 사용자는 서버의 문제로 판단할 여지가 있을 것 같은데요~ 

로직에 따라 다른 상태 코드나 이유 등을 내려주는 경우가 있을 것 같습니다. 이런 요구사항을 대응하기 위해서 이러한 Exception을 Handling 하는 객체는 무엇이고,  어떻게 적용할 수 있는지 찾아보시고 반영해주시면 좋을 것 같아요

tpleehan replied:
해당 부분은 try-catch를 통해 Exception을 Handling 하는 것이 맞는지 궁금합니다!
아니면 요구사항을 대응하기 위해 Custom Exception을 이용해서 예외를 정의하는 것이 좋은 방법인지 궁금합니다!!

brody-0125 replied:
> 해당 Exception이 발생하면 프론트 개발자나 사용자는 서버의 문제로 판단할 여지가 있을 것 같은데요~

해당 내용의 의미는 사용자에게 "이미 존재하는 아이디이다" 라는 상황을 인지시켜줄 필요가 있다라는 의미였습니다.
현재는 그런 정보까지 사용자에게 노출되는 것이 아니라 500이나 에러 페이지만 내려갈 것 같거든요!

brody-0125 commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -0,0 +1,85 @@
+				.email(requestDto.getEmail())
+				.authToken(UUID.randomUUID().toString())
+				.expired(false)
+				.build());
```
어떤 메서드의 인자로 Builder 생성 로직을 넣는건 가독성 상 좋지 않은 것 같습니다.

아래 Member도 포함입니다~

tpleehan replied:
해당 부분은 다시 찾아보고 알아보겠습니다😥

tpleehan committed:
chore: git cache reset

- git rm --cached 파일 삭제로 recommit

tpleehan committed:
chore: git cache reset

- git rm --cached 파일 삭제로 recommit

tpleehan committed:
Merge branch 'feature/6-signup-member' of https://github.com/E-TF/LaLa into feature/6-signup-member

tpleehan committed:
chore: CRLF, No newline at end of file

- Windows CRLF 개행 문자로 설정된 부분 LF로 변경
- No newline at end of file 변경

tpleehan committed:
chore: CRLF, No newline at end of file

- Windows CRLF 개행 문자로 설정된 부분 LF로 변경
- No newline at end of file 변경

tpleehan committed:
feat: #6 회원가입 (member)
- pr feedback 수정

brody-0125 commented on src/test/java/com/project/lala/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,57 @@
+			.name("test_name")
+			.email("test_email@email.test")
+			.build();
+	}
```
```
 	@BeforeEach
 	void beforeEach() {
 		member = Member.builder()
 			.loginId("test_id")
 			.password("!@#test123")
 			.nickname("test_nick")
 			.name("test_name")
 			.email("test_email@email.test")
 			.build();
 	}
```

현재 이 사이클은 사용되지 않는 것 같네요. 

이러한 초기 데이터 설정이 필요할 때 추가하는 것이 어떨까요?

tpleehan committed:
chore: CRLF, No newline at end of file

- Windows CRLF 개행 문자로 설정된 부분 LF로 변경
- No newline at end of file 변경

brody-0125 commented on src/test/java/com/project/lala/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,57 @@
+		Optional<Member> findMemberId = memberRepository.findById(joinMember.getId());
+
+		assertThat(findMemberId.get().getId()).isEqualTo(joinMember.getId());
+		assertThat(findMemberId.get().getEmail()).isEqualTo(joinMember.getEmail());
```
```java
	@Test
 	@Transactional
 	@Rollback(false)
 	public void signUp() {
 		Member joinMember = Member.createMember("test_id", encryptionService.encrypt("!@#test123"),
 			"test_nick", "test_name", "test_email@email.test");

 		Member signupMember = memberRepository.save(joinMember);
 		Optional<Member> findMemberId = memberRepository.findById(joinMember.getId());

 		assertThat(findMemberId.get().getId()).isEqualTo(joinMember.getId());
 		assertThat(findMemberId.get().getEmail()).isEqualTo(joinMember.getEmail());
```

이 부분은 오류를 해결하던 시점에 말씀을 따로 드렸던 것 같은데요~ 
테스트 하고자 하는 컴포넌트는 memberService라고 되어있지만 실제로는 memberService에 작성된 코드를 테스트 하는 것이 아니라 repository에 상태가 정상적으로 반영되었는가? 라는 memberRepository에 대한 테스트를 작성하고 계신 것 같습니다.

layer test와 unit test에 대해서 찾아보시면 좋을 것 같네요.
  * 테스트 코드의 목적은 무엇일까요?
    * 무엇을 테스트하고자 하는 것일까요?
    * 외부 의존성을 테스트하려면 어떻게 해야할까요?
    * SpringBootTest의 단점은 무엇이고 이게 심화되었을 때 어떤 상황이 발생할 수 있을까요?

tpleehan committed:
chore: CRLF, No newline at end of file

- Windows CRLF 개행 문자로 설정된 부분 LF로 변경
- No newline at end of file 변경

tpleehan committed:
fix: #6 회원가입(member)

- PR feedback 수정

brody-0125 commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -63,10 +63,10 @@ public SignUpResponseDto signUp(SignUpRequestDto requestDto) {
-		Member member = memberRepository.findByEmail(signUpRequestDto.getEmail())
+	public void confirmEmail(SignUpRequest signUpRequest) {
+		EmailAuth emailAuth = emailAuthRepository.findValidAuthByEmail(signUpRequest.getEmail(),
+			signUpRequest.getAuthToken(), LocalDateTime.now()).orElseThrow(IllegalArgumentException::new);
```
발급했던 인증 코드를 조회하는 로직이 현재는 Member Service 내부에만 있지만, 추후에 다른 기능을 제공해야 할 때에도 사용되어야 한다면 중복될 여지가 있을 수 있습니다. 

(+ 만약 메일 말고도 다른 방식으로 (카카오 알림톡) 인증을 제공하게 된다면, 여기에 코드나 분기가 추가될 여지가 있을 것 같구요.)

어떻게 하면 해당 로직을 분리해낼 수 있을지 고민하고 저에게 말씀 주시면 좋을 것 같습니다.
- Hint - 레이어 아키텍처에서 상위 레이어를 구성 및 참조하는 방법을 활용하면 될 것 같아요

brody-0125 commented on src/main/java/com/project/lala/service/MemberService.java:
```diff
@@ -52,7 +52,7 @@ public SignUpResponseDto signUp(SignUpRequestDto requestDto) {
 
 		emailService.sendEmail(emailAuth.getEmail(), emailAuth.getAuthToken());
```
Member Service에서 `emailService.sendEmail(emailAuth.getEmail(), emailAuth.getAuthToken());` 로직을 호출하는 부분이 눈에 보이는데요. 

저희가 사용하고 있는 레이어 아키텍처 원칙에서는 상위 레이어가 하위 레이어로 요청을 보내는 구조를 유지해야 한다는 느낌을 줍니다. (이 부분은 같은 레이어가 서로 참조하는 형태라고 볼 수 있겠죠.)

이런 구조가 좀 더 심화되었을 때 (Email Service에서도 사용자의 추가 정보 또는 상태를 알기 위해 Member Service를 참조하는 경우) 구조적으로 무슨 문제가 생길 수 있을 지 생각해보시면 좋을 것 같고, 아래의 confirmEmail 부분과 동일한 고민을 해주시면 좋을 것 같아요.

brody-0125 replied:
추가적으로 해당 Method는 Async가 적용되어 있는 것 같은데요. signUp Method가 처리 완료되었을 때 (Return 되었을 때) sendEmail Method가 실패하게 된다면 무슨 일이 발생할 수 있을지 고민해주시면 좋을 것 같아요.

brody-0125 commented on src/main/java/com/project/lala/controller/SignController.java:
```diff
@@ -17,22 +17,21 @@
+@RequestMapping("/sign")
 @RequiredArgsConstructor
-public class MemberController {
+public class SignController {
```
sign-in, sign-up 등은 서버에서 관리하는 Resource가 아닌 기능 단위, 즉 행위로 보아야할 단어인데요.

이게 URL에 포함되어 있는 것에 대한 고민이 필요할 것 같습니다. 

Controller가 SignController라고 해서 굳이 sign이라는 url을 사용할 필요는 없거든요.

tpleehan committed:
chore: CRLF, No newline at end of file
- Windows CRLF 개행 문자로 설정된 부분 LF로 변경
- No newline at end of file 변경

tpleehan committed:
Revert "chore: CRLF, No newline at end of file"

This reverts commit 5cf7c02cab33d408450a027cd3b009d620991353.

tpleehan committed:
chore: CRLF, No newline at end of file
- Windows CRLF 개행 문자로 설정된 부분 LF로 변경
- No newline at end of file 변경

tpleehan committed:
Refactor: for loop enhanced for 변경
- byte[] 배열 타입 enhanced for loop로 변경

tpleehan committed:
Test: 테스트 코드 추가
- Member Register 관련 Tdd 추가

tpleehan committed:
fix: #6 회원가입(member)
- PR feedback 수정(@Transactional annotation, legibility)

tpleehan committed:
fix: #6 회원가입(member)
- Mysql 8.0.17부터 Keyword(예약어) MEMBER 추가로 Table name MEMBERS 변경
 참고: https://dev.mysql.com/doc/mysqld-version-reference/en/keywords-8-0.html#keywords-8-0-detailed-M

tpleehan commented on src/main/java/com/project/lala/entity/Member.java:
```diff
@@ -0,0 +1,104 @@
+import lombok.NoArgsConstructor;
+
+@Entity
+@Table(name = "MEMBERS")
```
Mysql 8.0.17부터 Keyword(예약어) MEMBER 추가로 Table name MEMBERS 변경 

참고: https://dev.mysql.com/doc/mysqld-version-reference/en/keywords-8-0.html#keywords-8-0-detailed-M

tpleehan committed:
Test: method name, Test code 수정
- improvised method name 변경
- 완성되지 않은 Test code 제거

tpleehan committed:
fix: #6 회원가입(member)
- PR feed back 수정
- Custom Exception 추가

tpleehan committed:
Test: 테스트 코드 수정 및 추가 #6
- Sign Controller 테스트 추가

brody-0125 commented on src/test/java/com/project/lala/controller/SignControllerTest.java:
```diff
@@ -0,0 +1,59 @@
+import com.project.lala.dto.SignUpResponse;
+import com.project.lala.service.MemberService;
+
+@SpringBootTest
```
@SpringBootTest는 모든 Bean을 테스트를 수행하기 전에 Container에 모두 적재하게 되는데요. 그러다 보면 테스트 하고자하는 기능에서 필요하지 않는 의존성이 많은 상태가 되고, ApplicationContext를 띄우고 테스트하는데 걸리는 시간이 서비스 규모에 따라 크게 증가할 수 있습니다. 이 이슈는 실제 CI Pipeline에서 두드러지게 나타나기도 하구요.

가능하다면 각 Layer에 대한 테스트들은 WebMvcTest, DataJpaTest (Service의 경우 Test Configuration 활용) 등을 이용하는 것이 테스트 시간을 줄여줄 수 있습니다.

brody-0125 replied:
ex) ApplicationContext를 띄우는데 걸리는 시간이 30초고, @SpringBootTest를 사용하는 통합, 계층 테스트가 5개라면 띄우는 데에만 2분 30초가 걸릴 수 있습니다. [실제로는 이것보다 더 무거운 서비스가 많습니다.]

그럼 위의 케이스를 기준으로 생각해보았을 때 단위 테스트만을 가진 서비스보다 Docker Image를 만들어 통합하는데 걸리는 시간이 2분 30초 정도 더 걸린다는 거겠죠? 

사내에서 주기적으로 배포되는 프로젝트 수가 많아지고 소프트웨어 통합 주기가 빨라진다면 해당 서비스로 인해 CI Pipeline에서 생각보다 큰 지연이 발생하고 전체 프로젝트에 악영향을 미칠 수도 있는겁니다.

tpleehan committed:
fix: PR feedback 수정 #6
- layer architecture & service level refactoring

tpleehan committed:
fix: PR feedback 수정 #6
- feedback 수정 및 refactoring

tpleehan committed:
fix: PR feedback 수정 #6
- SpringBootTest -> WebMvcTest 변경
- EmailService 사용하지 않는 필드 제거

tpleehan committed:
test: 이메일 confirm 추가
- confirmEmail 테스트 코드 추가 및 method name 변경

tpleehan committed:
test: MemberService 테스트 수정
- memberRepository에 대한 테스트 코드를 Mockito를 이용해서 service 테스트로 변경
- SignUpRequest Builder 추가, 생성자 수정

tpleehan committed:
test: SignController 테스트 수정
- SignController Parameter 변경 및 테스트 코드 수정

brody-0125 commented on src/test/java/com/project/lala/service/MemberServiceTest.java:
```diff
@@ -1,71 +1,108 @@
+		verify(memberRepository, times(1)).findByEmail(request.getEmail());
+		verify(emailAuthRepository, times(1)).save(any());
+		verify(memberRepository, times(1)).save(any());
+		verify(emailService, times(1)).sendEmail(eq(request.getEmail()), anyString());
```
보낼 값이나 응답이 와야한다고 Stubbing하는 부분, 그리고 실제 검증하는 영역이 명시적으로 구분이 되어 있으면 좋을 것 같습니다.

brody-0125 commented on src/test/java/com/project/lala/service/MemberServiceTest.java:
```diff
@@ -1,71 +1,108 @@
-		Member joinMember = Member.createMember("test_id", encryptionService.encrypt("!@#test123"),
-			"test_nick", "test_name", "test_email@email.test");
+	@Mock
+	private EmailAuthRepository emailAuthRepository = Mockito.mock(EmailAuthRepository.class);
```
`Mockito.mock(EmailAuthRepository.class);`을 쓰셨다면 @Mock을 굳이 쓸 필요는 없을 것 같습니다.

tpleehan committed:
test: PR feedback 수정 #6
- Class AccessLevel, Builder 및 MemberService 테스트 코드 수정

tpleehan committed:
fix: Entity method 수정
- 해당 부분이 옳지 않은 코드이기 때문에 수정

tpleehan committed:
build: jpa naming 전략 변경
- camelCase -> snake_case
- Hibernate 4에 사용되는 strategy 변경(현 프로젝트는 Hibernate 5 버전 사용)

tpleehan commented on src/main/resources/application.yml:
```diff
@@ -4,8 +4,7 @@ spring:
       naming:
-        strategy: org.hibernate.cfg.ImprovedNamingStrategy
-        physical-strategy: org.hibernate.boot.model.naming.PhysicalNamingStrategyStandardImpl
+        physical-strategy: org.hibernate.boot.model.naming.CamelCaseToUnderscoresNamingStrategy
```
Hibernate 5 버전을 사용하고 있는데 4 버전에서 사용되는 hibernate.strategy를 수정합니다.
JPA Naming 전략 중 `SpringPhysicalNamingStrategy`은 Deprecate가 되어 `CamelCaseToUnderscoresNamingStrategy`로 사용합니다.

참고
- https://docs.spring.io/spring-boot/docs/2.6.0-M1/api/deprecated-list.html
- https://in.relation.to/2021/07/19/hibernate-orm-554-release/

tpleehan committed:
fix: Lombok @Table 어노테이션 제거
- jpa naming 전략 수정으로 불필요한 Lombok 제거

tpleehan committed:
Merge branch 'develop' into feature/6-signup-member

brody-0125 commented:
확인했습니다~ 머지 해주셔도 될 것 같아요 ㅎㅎ 머지 타입은 Squash Merge로 해주세요!
